README
NAME
    Alien::FLTK2 - Build and use the experimental 2.0.x branch Fast Light
    Toolkit

Description
    This distribution builds and installs libraries for the (experimental)
    `2.0.x' branch of the FLTK GUI toolkit.

Synopsis
        use Alien::FLTK2;
        use ExtUtils::CBuilder;
        my $AF     = Alien::FLTK2->new();
        my $CC     = ExtUtils::CBuilder->new();
        my $source = 'hello_world.cxx';
        open(my $FH, '>', $source) || die '...';
        syswrite($FH, <<'') || die '...'; close $FH;
          #include <fltk/Window.h>
          #include <fltk/Widget.h>
          #include <fltk/run.h>
          using namespace fltk;
          int main(int argc, char **argv) {
            Window *window = new Window(300, 180);
            window->begin();
            Widget *box = new Widget(20, 40, 260, 100, "Hello, World!");
            box->box(UP_BOX);
            box->labelfont(HELVETICA_BOLD_ITALIC);
            box->labelsize(36);
            box->labeltype(SHADOW_LABEL);
            window->end();
            window->show(argc, argv);
            return run();
          }

        my $obj = $CC->compile('C++'                => 1,
                               source               => $source,
                               include_dirs         => [$AF->include_dirs()],
                               extra_compiler_flags => $AF->cxxflags()
        );
        my $exe = $CC->link_executable(objects            => $obj,
                                       extra_linker_flags => $AF->ldflags());
        print system('./' . $exe) ? 'Aww...' : 'Yay!';
        END { unlink grep defined, $source, $obj, $exe; }

Constructor
    There are no per-object configuration options as of this version, but
    there may be in the future, so any new code using Alien::FLTK2 should
    create objects with the `new' constructor.

        my $AF = Alien::FLTK2->new( );

Methods
    After creating a new Alien::FLTK2 object, use the following methods to
    gather information:

  `include_dirs'
        my @include_dirs = $AF->include_dirs( );

    Returns a list of the locations of the headers installed during the
    build process and those required for compilation.

  `library_path'
        my $lib_path = $AF->library_path( );

    Returns the location of the private libraries we made and installed
    during the build process.

  `cflags'
        my $cflags = $AF->cflags( );

    Returns additional C compiler flags to be used.

  `cxxflags'
        my $cxxflags = $AF->cxxflags( );

    Returns additional flags to be used to when compiling C++ using FLTK.

  `ldflags'
        my $ldflags = $AF->ldflags( qw[gl images] );

    Returns additional linker flags to be used. This method can
    automatically add appropriate flags based on how you plan on linking to
    fltk. Acceptable arguments are:

    `static'
        Returns flags to link against a static FLTK library.

        FLTK's license allows static linking, but Alien::FLTK2 does not
        build static libs. ...yet.

    `gl'
        Include flags to use GL.

        *This is an experimental option. Depending on your system, this may
        also include OpenGL or MesaGL.*

    `images'
        Include flags to use extra image formats (PNG, JPEG).

  `branch'
        my $revision = $AF->branch( );

    Returns the SVN branch of the source Alien::FLTK2 was built with. For
    the `1.3.x' branch of fltk, see Alien::FLTK.

  `revision'
        my $revision = $AF->revision( );

    Returns the SVN revision number of the source Alien::FLTK2 was built
    with.

  `capabilities'
        my $caps = $AF->capabilities( );

    Returns a list of capabilities supported by your Alien::FLTK2
    installation. This list can be handed directly to `ldflags( )'.

  `config'
        my $configuration = $AF->config( );

    Returns a hashref containing the raw configuration data collected during
    build. This would be helpful when reporting bugs, etc.

Notes
  Requirements
    Prerequisites differ by system...

    Win32
        The fltk2 libs and Alien::FLTK2 both build right out of the box with
        MinGW. Further testing is needed for other setups.

    X11/*nix
        X11-based systems require several development packages. On Debian,
        these may be installed with:

            > sudo apt-get install libx11-dev
            > sudo apt-get install libxi-dev

        Additionally, the optional XCurser lib may be installed with:

            > sudo apt-get install libxcursor-dev

    Darwin/OSX
        Uh, yeah, I have no idea.

  Installation
    The distribution is based on Module::Build, so use the following
    procedure:

      > perl Build.PL
      > ./Build
      > ./Build test
      > ./Build install

  Support Links
    * Issue Tracker
        http://github.com/sanko/alien-fltk/issues

        Please only report Alien::FLTK2 related bugs to this tracker. For
        FLTK issues, use http://github.com/sanko/fltk-perl2/issues/

    * Commit Log
        http://github.com/sanko/alien-fltk/commits/master

    * Homepage:
        http://sanko.github.com/fltk-perl/ is the homepage of the FLTK
        project.

    * License:
        http://www.perlfoundation.org/artistic_license_2_0

        See the License and Legal section of this document.

    * Mailing List
        Once I find someone to host a list for the FLTK project, I'll use it
        for Alien::FLTK2 too.

    * Repository
        http://github.com/sanko/alien-fltk2/ and you are invited to fork it.

  Examples
    Please see the Synopsis and the files in the `/examples/'.

  Bugs
    Numerous, I'm sure.

  To Do
    Please see Alien::FLTK2::Todo

See Also
    FLTK, Alien::FLTK

Acknowledgments
    The FLTK Team - http://www.fltk.org/

Author
    Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

    CPAN ID: SANKO

License and Legal
    Copyright (C) 2009-2011 by Sanko Robinson <sanko@cpan.org>

    This program is free software; you can redistribute it and/or modify it
    under the terms of The Artistic License 2.0. See the LICENSE file
    included with this distribution or
    http://www.perlfoundation.org/artistic_license_2_0. For clarification,
    see http://www.perlfoundation.org/artistic_2_0_notes.

    When separated from the distribution, all POD documentation is covered
    by the Creative Commons Attribution-Share Alike 3.0 License. See
    http://creativecommons.org/licenses/by-sa/3.0/us/legalcode. For
    clarification, see http://creativecommons.org/licenses/by-sa/3.0/us/.

    Alien::FLTK2 is based in part on the work of the FLTK project. See
    http://www.fltk.org/.



Changes
Version 0.09296 | 2012-01-24 17:46:38Z (Tue, 24 Jan 2012) | df17165

    Upstream Changes:
     * [r9296] Fix inconsistent events sent when pressing and releasing
         multiple keys on X11

Version 0.09204 | 2011-04-02 02:11:40Z (Sat, 02 Apr 2011) | ef7ae5b

    Upstream Changes:
     * [r8638] New built-in XP theme
     * [r9160] Keyboard support in ScrollGroup

Version 0.08550 | 2011-04-02 02:11:40Z (Sat, 02 Apr 2011) | ef7ae5b

    Upstream Changes:
     * [r8385] FileChooser now allows users to type absolute paths
     * [r8369] Improves utf8 in fltk::fltk_fopen()
     * [r8418] Upgrade libjpeg to latest version (as of 2011-02-13)
     * [r8416] Upgrade zlib to the latest version (as of 2010-12-02)
     * [r8414] Finishes upgrade of libpng
     * [r8465] FileIcon now loads some defaults under Gnome

    Notes:
     * Mirrored fltk2 snapshots renamed from fltk-2.0.x-rXXXX.tar.ext to
         fltk-2.0.x-alpha-rXXXX.tar.ext to make the unstable nature even more
         evident

Version 0.08365 | 2011-02-08 03:43:36Z (Tue, 08 Feb 2011) | ec4e842

    Upstream Changes:
     * [r8328, r8329, r8350] Improves FileChooser's Unicode support on Win32

Version 0.08323001 | 2011-02-02 23:58:04Z (Wed, 02 Feb 2011) | 8088b1a

    Resolved Issues/Bug Fixes:
     * Use YAML::Tiny directly when creating config during build

Version 0.08323 | 2011-02-02 21:25:09Z (Wed, 02 Feb 2011) | 7285db6

    Upstream Changes:
     * [r7723] Fixed STR #2431. Clipping region fix for fltk::TiledImage
     * [r8317] Added in horizontal mouse scrolling for windows and resolves
          STR #2325 and half of #1637 by including dirent.h on mingw compilers
     * [r8316] Patch for FileChooser that allows it to keep the "suggested"
          filename (STR #2488 and STR #1702) when both navigating and to begin
          with.
     * [r8313] Added in horizontal scrolling for Linux (STR #1964).
     * [r8312] Fix to allow shared libraries to build under msys (STR #2487).
     * [r8304] Added in a SONAME for shared objects (STR #2109) to aid
          side-by-side 1.x and 2.x installations.
     * [r8302] Committe a patch for STRs #2493 and #2318 regarding multiple
          keyboard layouts and X.

    Notes:
     * select patch for win32 updated to apply cleanly against r8323

Version 0.07725001 | 2011-01-10 04:59:34Z (Mon, 10 Jan 2011) | 0546209

    Resolved Issues/Bug Fixes:
     * Add patch for "dynamic" select timeout on win32

Version 0.07725 | 2011-01-10 04:59:34Z (Mon, 10 Jan 2011) | 0546209

    Resolved Issues/Bug Fixes:
     * Disable (broken) async fh select on win32

    Upstream Changes
     * [r7723] Fixed STR #2431. Clipping region fix for fltk::TiledImage

    Notes:
     * Simple patch system added to build cycle

Version 0.07704001 | 2010-09-25 14:04:33Z (Sat, 25 Sep 2010) | c262431

    Resolved Issues/Bug Fixes:
     * #define symbols for dirent location were malformed

Version 0.07704 | 2010-09-12 03:22:02Z (Sun, 12 Sep 2010) | 68973d4

    Resolved Issues/Bug Fixes:
     * Possible undefined fix for Solaris. Eh, stranger things have happend...

    Upstream Changes:
     * [r7703] Fixed STR #1902. (Critical bug under Win2k and VirtualBox)
     * [r7702] Updated README.windows (STR #2245). Thanks to rgarcia.
     * [r7698] Fixed double free of ValueSlider's label (STR #2412)

Version 0.07680002 | 2010-09-12 03:22:02Z (Sun, 12 Sep 2010) | 68973d4

    Resolved Issues/Bug Fixes:
     * Correcting c+p bug in gl test
          See cpantesters.org/cpan/report/5856497a-ba06-11df-b6d3-36bb8fbefd91
     * When fltk2_gl disabled, do not attempt to build fltk2_glut lib either
     * Work around a bug in v5.12.1 of 32bit Strawberry Perl which places some
          headers under ...strawberry\c\i686-w64-mingw32\include\ without
          adding the directory to $Config{incpath} or even $Config{userinc}.
          This was causing such systems to miss GL during configure.
     * Build against fltk2's mainline SVN revision 7680
     * Improved (I hope) library archiver search for MinGW-based systems

Version 0.06970024 | 2010-09-04 04:21:58Z (Sat, 04 Sep 2010) | 978608d

    Resolved Issues/Bug Fixes:
     * Undertested GL build fix
     * ExtUtils::CBuilder tacks perl's CORE include dir on when compiling so
          the config.h we generate for FLTK2 was left unused and caused
          problems w/MinGW
          See cpantesters.org/cpan/report/07665931-b19f-3f77-b713-d32bba55d77f

    Documentation/Sample Code/Test Suite:
     * New GL test

Version 0.06970023 | 2010-04-18 19:40:15Z (Sun, 18 Apr 2010) | b154909

    Resolved Issues/Bug Fixes:
     * Untested config fix for Solaris
          See STR #1890
          See cpantesters.org/cpan/report/07134937-b19f-3f77-b713-d32bba55d77f
     * Untested config fix for OSX
          See cpantesters.org/cpan/report/07134608-b19f-3f77-b713-d32bba55d77f

Version 0.06970022 | 2010-03-29 05:13:06Z (Mon, 29 Mar 2010) | a8a2473

    Resolved Issues/Bug Fixes:
     * Untested config step for OSX

Version 0.06970021 | 2010-03-29 05:13:06Z (Mon, 29 Mar 2010) | a8a2473

    Resolved Issues/Bug Fixes:
     * [7f6c18b] Font sizing on Win32 w/unicode
           See http://www.cpantesters.org/cpan/report/7045261

Version 0.06970020 | 2010-03-29 05:13:06Z (Mon, 29 Mar 2010) | a8a2473

    Resolved Issues/Bug Fixes:
     * Under-tested MSVC fixes in the fltk2 lib

Version 0.06970019 | 2010-02-15 21:58:16Z (Mon, 15 Feb 2010) | 29adc40

    Resolved Issues/Bug Fixes:
     * Un(der)tested CygWin configuration and build fixes

    Documentation/Sample Code/Test Suite:
     * examples/0001_synopsis.pl had a fail-worthy typo

    Notes:
     * For now, we only support Win32-ish CygWin builds. In the future, we may
       include X11-like builds. But don't count on it.

Version 0.06970018 | 2010-02-13 16:55:35Z (Sat, 13 Feb 2010) | dee91f4

    Resolved Issues/Bug Fixes:
     * Rather than X11, OSX/Darwin uses Carbon for graphics so we no longer
       search for X11-related libs


lib/Alien/FLTK.pm
package Alien::FLTK;
{
    use strict;
    use warnings;
    use File::Spec::Functions qw[catdir rel2abs canonpath];
    our $BASE = 0; our $SVN = 6985; our $DEV = 0; our $VERSION = sprintf('%d.%05d' . ($DEV ? '_%03d' : ''), $BASE, $SVN, $DEV);
    my $_config = eval do { local $/; <DATA> }
        or warn
        "Couldn't load Alien::FLTK configuration data: $@\n Using defaults";
    close DATA;
    sub new { return bless \$|, shift; }
    sub config   { return $_config; }
    sub revision { return $SVN; }
    sub branch   { return $_config->{'branch'} }

    sub include_dirs {
        my ($self) = @_;
        my @return = keys %{
              $self->config->{'include_dirs'}
            ? $self->config->{'include_dirs'}
            : ()
            };
        for my $path (catdir(qw[.. .. blib arch Alien FLTK]),
                      catdir(qw[. blib arch Alien FLTK]),
                      catdir(qw[Alien FLTK]))
        {   foreach my $inc (@INC) {
                next unless defined $inc and !ref $inc;
                my $dir = rel2abs(
                     catdir($inc, $path, 'include', 'fltk-' . $self->branch));
                return ($dir, @return) if -d $dir && -r $dir;
            }
        }
        return undef;
    }

    sub library_path {
        my ($self) = @_;
        for my $path (catdir(qw[.. .. blib arch Alien FLTK]),
                      catdir(qw[. blib arch Alien FLTK]),
                      catdir(qw[Alien FLTK]))
        {   foreach my $inc (@INC) {
                next unless defined $inc and !ref $inc;
                my $dir = rel2abs(
                        catdir($inc, $path, 'libs', 'fltk-' . $self->branch));
                return $dir if -d $dir && -r $dir;
            }
        }
        return undef;
    }
    sub cflags { return shift->cxxflags(); }

    sub cxxflags {
        my ($self) = @_;
        return $self->config->{'cxxflags'} ? $self->config->{'cxxflags'} : '';
    }

    sub ldflags {    # XXX - Cache this
        my ($self, @args) = @_;

        #
        my $libdir = shift->library_path();

        # Calculate needed libraries
        my $SHAREDSUFFIX
            = $self->config->{'_a'} ? $self->config->{'_a'}
            : $^O =~ '$MSWin32' ? '.a'
            :                     '.o';
        my $LDSTATIC = sprintf '-L%s %s/libfltk%s%s %s',
            $libdir, $libdir, ($self->branch eq '1.3.x' ? '' : '2'),
            $SHAREDSUFFIX,
            ($self->config->{'ldflags'} ? $self->config->{'ldflags'} : '');
        my $LDFLAGS = "-L$libdir "
            . ($self->config->{'ldflags'} ? $self->config->{'ldflags'} : '');
        my $LIBS = sprintf '%s/libfltk%s%s', $libdir,
            ($self->branch eq '1.3.x' ? '' : '2'),
            $SHAREDSUFFIX;
        if (grep {m[forms]} @args) {
            $LDFLAGS = sprintf '-lfltk%s_forms %s',
                ($self->branch eq '1.3.x' ? '' : '2'), $LDFLAGS;
            $LDSTATIC = sprintf '$libdir/libfltk%s_forms%s %s',
                $libdir, ($self->branch eq '1.3.x' ? '' : '2'), $SHAREDSUFFIX,
                $$LDSTATIC;
            $LIBS = sprintf '%s %s/libfltk%s_forms%s',
                $LIBS, $libdir, ($self->branch eq '1.3.x' ? '' : '2'),
                $SHAREDSUFFIX;
        }
        if ((grep {m[gl]} @args) && $self->config->{'GL'}) {
            my $LIBGL = $self->config->{'GL'};
            $LDFLAGS = sprintf '-lfltk%s_gl %s %s',
                ($self->branch eq '1.3.x' ? '' : '2'),
                $LIBGL, $LDFLAGS;
            $LDSTATIC = sprintf '%s/libfltk%s_gl%s %s %s',
                $libdir, ($self->branch eq '1.3.x' ? '' : '2'),
                $SHAREDSUFFIX, $LIBGL, $LDSTATIC;
            $LIBS = sprintf '%s %s/libfltk%s_gl%s',
                $LIBS, $libdir,
                ($self->branch eq '1.3.x' ? '' : '2'),
                $SHAREDSUFFIX;
        }
        if ((grep {m[images]} @args) && $self->config->{'ldflags_image'}) {
            $LDFLAGS  = $self->config->{'ldflags_image'} . " $LDFLAGS";
            $LDSTATIC = sprintf '%s/libfltk%s_images%s %s %s',
                $libdir, ($self->branch eq '1.3.x' ? '' : '2'),
                $SHAREDSUFFIX, $LDSTATIC, $self->config->{'ldflags_image'};
        }
        return (
             ((grep {m[static]} @args) ? $LDSTATIC : $LDFLAGS) . ' -lsupc++');
    }

    sub capabilities {
        my ($self) = @_;
        my @caps;
        push @caps, 'gl' if $self->config->{'config'}{'HAVE_GL'};

        # TODO: images, forms, static(?)
        return @caps;
    }
    1
}

=pod

=head1 NAME

Alien::FLTK - Build and use the Fast Light Toolkit binaries

=head1 Description

This distribution builds and installs libraries for the (experimental)
C<2.0.x> branch of the FLTK GUI toolkit.

=head1 Synopsis

    use Alien::FLTK;
    use ExtUtils::CBuilder;
    my $AF     = Alien::FLTK->new();
    my $CC     = ExtUtils::CBuilder->new();
    my $source = 'hello_world.cxx';
    open(my $FH, '>', $source) || die '...';
    syswrite($FH, <<'') || die '...'; close $FH;
      #include <fltk/Window.h>
      #include <fltk/Widget.h>
      #include <fltk/run.h>
      using namespace fltk;
      int main(int argc, char **argv) {
        Window *window = new Window(300, 180);
        window->begin();
        Widget *box = new Widget(20, 40, 260, 100, "Hello, World!");
        box->box(UP_BOX);
        box->labelfont(HELVETICA_BOLD_ITALIC);
        box->labelsize(36);
        box->labeltype(SHADOW_LABEL);
        window->end();
        window->show(argc, argv);
        return run();
      }

    my $obj = $CC->compile('C++'                => 1,
                           source               => $source,
                           include_dirs         => [$AF->include_dirs()],
                           extra_compiler_flags => $AF->cxxflags()
    );
    my $exe = $CC->link_executable(objects            => $obj,
                                   extra_linker_flags => $AF->ldflags());
    print system('./' . $exe) ? 'Aww...' : 'Yay!';
    END { unlink grep defined, $source, $obj, $exe; }

=head1 Constructor

There are no per-object configuration options as of this version, but there
may be in the future, so any new code using L<Alien::FLTK|Alien::FLTK> should
create objects with the C<new> constructor.

    my $AF = Alien::FLTK->new( );

=head1 Methods

After creating a new L<Alien::FLTK|Alien::FLTK> object, use the following
methods to gather information:

=head2 C<include_dirs>

    my @include_dirs = $AF->include_dirs( );

Returns a list of the locations of the headers installed during the build
process and those required for compilation.

=head2 C<library_path>

    my $lib_path = $AF->library_path( );

Returns the location of the private libraries we made and installed
during the build process.

=head2 C<cflags>

    my $cflags = $AF->cflags( );

Returns additional C compiler flags to be used.

=head2 C<cxxflags>

    my $cxxflags = $AF->cxxflags( );

Returns additional flags to be used to when compiling C++ using FLTK.

=head2 C<ldflags>

    my $ldflags = $AF->ldflags( qw[gl images] );

Returns additional linker flags to be used. This method can automatically add
appropriate flags based on how you plan on linking to fltk. Acceptable
arguments are:

=over

=item C<static>

Returns flags to link against a static FLTK library.

FLTK's license allows static linking, but L<Alien::FLTK|Alien::FLTK> does not
build static libs. ...yet.

=item C<gl>

Include flags to use GL.

I<This is an experimental option. Depending on your system, this may also
include OpenGL or MesaGL.>

=item C<images>

Include flags to use extra image formats (PNG, JPEG).

=begin TODO

=item C<glut>

Include flags to use FLTK's glut compatibility layer.

=item C<forms>

Include flags to use FLTK's forms compatibility layer.

=end TODO

=back

=head2 C<branch>

    my $revision = $AF->branch( );

Returns the SVN brance of the source L<Alien::FLTK|Alien::FLTK> was built
with.

Currently, L<Alien::FLTK|Alien::FLTK> defaults to the 2.0.x branch but it is
capable of building the more stable 1.3.x branch.

=head2 C<revision>

    my $revision = $AF->revision( );

Returns the SVN revision number of the source L<Alien::FLTK|Alien::FLTK>
was built with.

=head2 C<capabilities>

    my $caps = $AF->capabilities( );

Returns a list of capabilities supported by your L<Alien::FLTK|Alien::FLTK>
installation. This list can be handed directly to
L<C<ldflags( )>|Alien::FLTK/ldflags>.

=head2 C<config>

    my $configuration = $AF->config( );

Returns a hashref containing the raw configuration data collected during
build. This would be helpful when reporting bugs, etc.

=head1 Notes

=head2 Requirements

Prerequisites differ by system...

=over

=item Win32

The fltk2 libs and L<Alien::FLTK|Alien::FLTK> both build right out of the box
with MinGW. Further testing is needed for other setups.

=item X11/*nix

X11-based systems require several development packages. On Debian, these may
be installed with...

  > sudo apt-get install libx11-dev
  > sudo apt-get install libxi-dev
  > sudo apt-get install libxcursor-dev

=item Darwin/OSX

Uh, yeah, I have no idea.

=back

=head2 Installation

The distribution is based on L<Module::Build|Module::Build>, so use the
following procedure:

  > perl Build.PL
  > ./Build
  > ./Build test
  > ./Build install

=head2 Support Links

=over

=item * Issue Tracker

http://github.com/sanko/alien-fltk/issues

Please only report L<Alien::FLTK|Alien::FLTK> related bugs to this tracker.
For L<FLTK|FLTK> issues, use http://github.com/sanko/fltk-perl/issues/

=item * Commit Log

http://github.com/sanko/alien-fltk/commits/master

=item * Homepage:

http://sanko.github.com/fltk-perl/ is the homepage of the L<FLTK|FLTK>
project.

=item * License:

http://www.perlfoundation.org/artistic_license_2_0

See the L<License and Legal|/"License and Legal"> section of this document.

=item * Mailing List

Once I find someone to host a list for the L<FLTK|FLTK> project, I'll use it
for L<Alien::FLTK|Alien::FLTK> too.

=item * Repository

http://github.com/sanko/alien-fltk/ and you are invited to fork it.

=back

=head2 Examples

Please see the L<Synopsis|/"Synopsis"> and the files in the C</examples/>.

=head2 Bugs

Numerous, I'm sure.

=head2 To Do

Please see L<Alien::FLTK::Todo|Alien::FLTK::Todo>

=head1 See Also

L<FLTK|FLTK>

=head1 Acknowledgments

=over

=item The FLTK Team - http://www.fltk.org/

=back

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2009 by Sanko Robinson E<lt>sanko@cpan.orgE<gt>

This program is free software; you can redistribute it and/or modify it under
the terms of The Artistic License 2.0. See the F<LICENSE> file included with
this distribution or http://www.perlfoundation.org/artistic_license_2_0.  For
clarification, see http://www.perlfoundation.org/artistic_2_0_notes.

When separated from the distribution, all POD documentation is covered by the
Creative Commons Attribution-Share Alike 3.0 License. See
http://creativecommons.org/licenses/by-sa/3.0/us/legalcode.  For
clarification, see http://creativecommons.org/licenses/by-sa/3.0/us/.

L<Alien::FLTK|Alien::FLTK> is based in part on the work of the FLTK project.
See http://www.fltk.org/.

=for git $Id$

=cut

__DATA__
do{ my $x = { }; $x; }
